eip-0027.md

```diff
@@ -0,0 +1,256 @@
+
+The checks can be switched off via a soft-fork (via voting on disabling rule #123).
+
+Contracts
+-------------------
+
+**Re-emission contract**: This contract pays 3 ERG per block starting from block #2,080,800 . Also, this contract allows for merging with other boxes; the merging transaction must have only two outputs; the first output is for the re-emission contract, the second one is to pay a mining fee (its value can be 0.01 ERG at most)
+
+```scala
+    val reemissionRewardPerBlock = monetarySettings.oneEpochReduction // 3 ERG
+
+    // output of the reemission contract
+    val reemissionOut = ByIndex(Outputs, IntConstant(0))
+
+    // output to pay miner
+    val minerOut = ByIndex(Outputs, IntConstant(1))
+
+    val rOutTokens = OptionGet(ExtractRegisterAs(reemissionOut, R2)(SCollection(STuple(SCollection(SByte), SLong))))
+
+    val firstTokenId = SelectField(ByIndex(rOutTokens, IntConstant(0)), 0.toByte)
+
+    val correctNftId = EQ(firstTokenId, ByteArrayConstant(reemissionNftId))
+    
+    // output of the reemission contract
+    val reemissionOut = ByIndex(Outputs, IntConstant(0))
+    
+    // output to pay miner
+    val minerOut = ByIndex(Outputs, IntConstant(1))
+    
+    // miner's output must have a script which is the time-locking reward for the miner's pubkey
+    // box height must be the same as block height
+    val correctMinerOutput = AND(
+      EQ(ExtractScriptBytes(minerOut), expectedMinerOutScriptBytesVal(monetarySettings.minerRewardDelay, MinerPubkey)),
+      EQ(Height, boxCreationHeight(minerOut))
+    )
+    
+    // reemission output's height must be the same as block height
+    val heightCorrect = EQ(boxCreationHeight(reemissionOut), Height)
+    
+    // reemission output's height is greater than reemission input
+    val heightIncreased = GT(Height, boxCreationHeight(Self))
+    
+    // check that height is greater than end of emission (>= 2,080,800 for the mainnet)
+    val afterEmission = GE(Height, IntConstant(emissionPeriod))
+    
+    // reemission contract must be preserved
+    val sameScriptRule = EQ(ExtractScriptBytes(Self), ExtractScriptBytes(reemissionOut))
+    
+    // miner's reward
+    val correctCoinsIssued = EQ(reemissionRewardPerBlock, Minus(ExtractAmount(Self), ExtractAmount(reemissionOut)))
+    
+    // when reemission contract box got merged with other boxes
+    val sponsored = AND(
+        GT(ExtractAmount(reemissionOut), ExtractAmount(Self)),
+        LE(ExtractAmount(ByIndex(Outputs, IntConstant(1))), LongConstant(10000000)), // 0.01 ERG
+        EQ(SizeOf(Outputs), 2)
+    )
+    
+    AND(
+        correctNftId,
+        sameScriptRule,
+        OR(
+          sponsored,
+          AND(
+            heightCorrect,
+            correctMinerOutput,
+            afterEmission,
+            heightIncreased,
+            correctCoinsIssued
```

Comment:
I think we also need to ensure one of two things 
A) INPUTS.size == 1
B) When miner spends the box, it always goes to proxy contract and not the re-emission contract. 

if we don't ensure both, then multiple re-emission contract boxes can be spent during reward collection

Reply:
There's only one re-emission contract box collecting rewards from emission contract (which is enforced by NFT condition in pay-to-reemission contract). 

```
INPUTS.size == 1
```

is not needed, better to leave possibility to have additional inputs (e.g. rewarding miners with custom tokens, if first input has re-emission contract box NFT).